yapapi/services.py

```diff
@@ -366,39 +388,73 @@ async def _run_instance(self, ctx: WorkContext):
                 (batch_task, signal_task), return_when=asyncio.FIRST_COMPLETED
             )
 
+            def change_state(stop=False, exc_info=(None, None, None)):
+                """Initiate state transition, possibly due to an error."""
+
+                nonlocal batch_task, handler, instance
+
+                if not stop and exc_info == (None, None, None):
+                    # Normal, i.e. non-error, state transition
+                    instance.service_state.lifecycle()
+                else:
+                    # Transition due to an error or `ControlSignal.stop`
+                    # TODO: define this transition just like Service.lifecycle()?
+                    # isn't it the same as `(stop | terminate)()`?
+                    if instance.state == ServiceState.running:
+                        instance.service_state.stop()
+                    else:
+                        instance.service_state.terminate()
+
+                handler = self._get_handler(instance)
+                logger.debug("%s state changed to %s", instance.service, instance.state.value)
+                instance.service._exc_info = exc_info
+                if batch_task:
+                    batch_task.cancel()
+                    # TODO: await batch_task here?
+                batch_task = None
+
             if batch_task in done:
                 # Process a batch
                 try:
+                    # Errors in service code will be raised by `batch_task.result()`.
+                    # These include:
+                    # - StopAsyncIteration's resulting from normal exit from handler function
+                    # - BatchErrors that were thrown into the service code by
+                    #   the `except BatchError` clause below
                     batch = batch_task.result()
-                    fut_result = yield batch
-                    result = await fut_result
-                    wrapped_results = loop.create_future()
-                    wrapped_results.set_result(result)
-                    batch_task = loop.create_task(handler.asend(wrapped_results))
                 except StopAsyncIteration:
-                    instance.service_state.lifecycle()
-                    handler = None
-                    batch_task = None
+                    change_state()
```

Comment:
@azawlocki wouldn't it be possible to make `change_state` a private (even if static) method on the `Cluster` ? 

(for the sake of testing etc)

Reply:
@shadeofblue `change_instance` is rather tightly integrated with the function in which it's defined (`Cluster._run_instance`) -- it uses its local variables `batch_tasks` and `handler`.

Some parts of `change_state` could be extracted to a separate method but I'm not sure if it's worth the hassle.

Reply:
@shadeofblue but you may decide for yourself, I did that in https://github.com/golemfactory/yapapi/pull/498/commits/dc741d6906a3bf285cf8280502203cb36da377a4